### ISSUE TITLE
Nested Container Tests not being run.

### ISSUE THREAD

**Issue by apotema** [ziglang/zig#18032]
### Zig Version

0.12.0-dev.1643+91329ce94

### Steps to Reproduce and Observed Behavior

zig test test.zig

It is supposed to say that all 5 tests passed, but it just says 4.
The nested test is not being run despite the documentaiton saying it would.

https://ziglang.org/documentation/master/#Nested-Container-Tests

test.zig file
```zig
const std = @import("std");
const expect = std.testing.expect;

// Imported source file tests will run when referenced from a top-level test declaration.
// The next line alone does not cause "testing_introduction.zig" tests to run.
const imported_file = @import("testing_introduction.zig");

test {
    // To run nested container tests, either, call `refAllDecls` which will
    // reference all declarations located in the given argument.
    // `@This()` is a builtin function that returns the innermost container it is called from.
    // In this example, the innermost container is this file (implicitly a struct).
    std.testing.refAllDecls(@This());

    // or, reference each container individually from a top-level test declaration.
    // The `_ = C;` syntax is a no-op reference to the identifier `C`.
    _ = S;
    _ = U;
    _ = @import("testing_introduction.zig");
}

const S = struct {
    test "S demo test" {
        try expect(true);
    }

    const SE = enum {
        V,

        // This test won't run because its container (SE) is not referenced.
        test "This Test Won't Run" {
            try expect(false);
        }
    };
};

const U = union { // U is referenced by the file's top-level test declaration
    s: US,        // and US is referenced here; therefore, "U.Us demo test" will run

    const US = struct {
        test "U.US demo test" {
            // This test is a top-level test declaration for the struct.
            // The struct is nested (declared) inside of a union.
            try expect(true);
        }
    };

    test "U demo test" {
        try expect(true);
    }
};
```

### Expected Behavior

It is supposed to say: `All 5 tests passed.`

**Comment by Vexu** [comment]
The documentation is outdated and should be using `refAllDeclsRecursive`.

**Comment by apotema** [comment]
My steps are

- Create a new folder
- Enter new folder
- `zig init-exe`
- Create file src/test.zig
- Copy the content from `https://ziglang.org/documentation/master/#Nested-Container-Tests` and put it into test.zig
- Remove everything related to `testing_introduction.zig`
- Run `zig test src/test.zig`

Result
`All 3 tests passed.`

It should be `All 4 tests passed`

I tried with `refAllDeclsRecursive` and had no success as well @Vexu .

**Comment by Vexu** [comment]
OK, I understand the issue now. `US` won't be referenced unless the fields of `U` are resolved which can be archived by for example using `@sizeOf(U)` instead of just `U`.

**Comment by Vexu** [comment]
`refAllDeclsRecursive` would also work if `U` was made `pub`, a requirement that did not yet exist when that example was last modified.

**Comment by apotema** [comment]
That solved the issue @Vexu .
Thanks

**Comment by dweiller** [comment]
> `refAllDeclsRecursive` would also work if `U` was made `pub`, a requirement that did not yet exist when that example was last modified.

Can you explain to me why `refAllDeclsRecursive` catches the test in `U.US` when `U` is `pub`, but `refAllDecls` doesn't? This seems like a wonky edge case that show my understanding of how referencing works is subtly off.

**Comment by Vexu** [comment]
The `pub` is required since both of the `refAllDelcs` functions are implemented using `@typeInfo` which only includes `pub` decls and only `refAllDeclsRecursive` references the test since it asks the compiler for `U`'s declarations.

**Comment by dweiller** [comment]
> The `pub` is required since both of the `refAllDelcs` functions are implemented using `@typeInfo` which only includes `pub` decls and only `refAllDeclsRecursive` references the test since it asks the compiler for `U`'s declarations.

I think I might have been a little unclear as this doesn't answer what I was thinking of. Currently if you have this file:

```zig
test {
    std.testing.refAllDeclsRecursive(@This());
}

pub const U = struct {
    field: US,

    const US = struct {
        test {}
    };
};

const std = @import("std");
```
and run `zig test` on it, both tests get picked up. If `std.testing.refAllDecls(@This());`, or `_ = U;` is used instead, then only the top level test is picked up. If `std.testing.refAllDecls(U);` used, then both tests are picked up.

I found this surprising since `US` is not `pub`. After more time to think I think I realise that since `field` has type `US` if anything (like `@typeInfo`) needs to analyze the fields of `U`, then `US` will get analyzed and the second test picked up.  I suppose this means that referencing a container does not mean that its fields get referenced.

**Comment by Vexu** [comment]
> I found this surprising since `US` is not `pub`.

`US` being `pub` or not is not relevant here since the compiler will scan all the declarations in `U` either way.

> After more time to think I think I realise that since `field` has type `US` if anything (like `@typeInfo`) needs to analyze the fields of `U`, then `US` will get analyzed and the second test picked up.

That's not the reason or either `std.testing.refAllDecls(@This());` would work too. `US` gets referenced by `refAllDeclsRecursive` because it sees `U` recurses into it and asks the compiler to scan for all its declarations.

**Comment by dweiller** [comment]
> That's not the reason or either `std.testing.refAllDecls(@This());` would work too.
This is not true, at least not for zig version 0.12.0-dev.1643+91329ce94 - using `std.testing.refAllDecls(@This());` only picks up the top level test.

**Comment by pfgithub** [comment]
In zig 0.13.0, this is outputting 3 tests passed, but in 0.14.0-dev.2443+3f7b3daaa it is outputting 4 tests passed so this seems to be fixed? The section has also seems to be gone from the documentation.

**Comment by mlugg** [comment]
Yes, this should now work as OP expected because referenced types are now always fully resolved.
